provider/provider.go

```diff
@@ -0,0 +1,1370 @@
+
+	s.connectivity.TriggerCheck()
+}
+
+func (s *SweepingProvider) batchProvide(prefix bitstr.Key, keys []mh.Multihash) {
+	if len(keys) == 0 {
+		return
+	}
+	addrInfo, ok := s.selfAddrInfo()
+	if !ok {
+		// Don't provide if the node doesn't have a valid address to include in the
+		// provider record.
+		return
+	}
+	if len(keys) <= individualProvideThreshold {
+		// Don't fully explore the region, execute simple DHT provides for these
+		// keys. It isn't worth it to fully explore a region for just a few keys.
+		s.individualProvide(prefix, keys, false, false)
+		return
+	}
+
+	regions, coveredPrefix, err := s.exploreSwarm(prefix)
+	if err != nil {
+		s.failedProvide(prefix, keys, fmt.Errorf("reprovide '%s': %w", prefix, err))
+		return
+	}
+	logger.Debugf("provide: requested prefix '%s' (len %d), prefix covered '%s' (len %d)", prefix, len(prefix), coveredPrefix, len(coveredPrefix))
+
+	// Add any key matching the covered prefix from the provide queue to the
+	// current provide batch.
+	extraKeys := s.provideQueue.DequeueMatching(coveredPrefix)
+	keys = append(keys, extraKeys...)
+	regions = keyspace.AssignKeysToRegions(regions, keys)
+
+	if !s.provideRegions(regions, addrInfo, false, false) {
+		logger.Errorf("failed to reprovide any region for prefix %s", prefix)
+	}
+}
+
+// selfAddrInfo returns the current peer.AddrInfo to be used in the provider
+// records sent to remote peers.
+//
+// If the node currently has no valid multiaddress, return an empty AddrInfo
+// and false.
+func (s *SweepingProvider) selfAddrInfo() (peer.AddrInfo, bool) {
+	addrs := s.getSelfAddrs()
+	if len(addrs) == 0 {
+		logger.Warn("provider: no self addresses available for providing keys")
+		return peer.AddrInfo{}, false
+	}
+	return peer.AddrInfo{ID: s.peerid, Addrs: addrs}, true
+}
+
+// individualProvide provides the keys sharing the same prefix to the network
+// without exploring the associated keyspace regions. It performs "normal" DHT
+// provides for the supplied keys, handles failures and schedules next
+// reprovide is necessary.
+func (s *SweepingProvider) individualProvide(prefix bitstr.Key, keys []mh.Multihash, reprovide bool, periodicReprovide bool) {
+	if len(keys) == 0 {
+		return
+	}
+
+	var provideErr error
+	if len(keys) == 1 {
+		coveredPrefix, err := s.vanillaProvide(keys[0])
+		if err == nil {
+			s.provideCounter.Add(context.Background(), 1)
+		} else if !reprovide {
+			// Put the key back in the provide queue.
+			s.failedProvide(prefix, keys, fmt.Errorf("individual provide failed for prefix '%s', %w", prefix, err))
+		}
+		provideErr = err
+		if periodicReprovide {
+			// Schedule next reprovide for the prefix that was actually covered by
+			// the GCP, otherwise we may schedule a reprovide for a prefix too short
+			// or too long.
+			s.reschedulePrefix(coveredPrefix)
+		}
+	} else {
+		wg := sync.WaitGroup{}
+		success := atomic.Bool{}
+		for _, key := range keys {
+			wg.Add(1)
+			go func() {
```

Comment:
If there are a huge number of key, this could create excessive goroutines. Maybe use a non-blocking worker pool to limit concurrency. Maybe https://github.com/gammazero/workerpool?tab=readme-ov-file#workerpool